### DISCUSSION TITLE
Multi threading in IfcOpenShell piecewise_function

### DISCUSSION THREAD

**Discussion by RickBrice** [IfcOpenShell/IfcOpenShell#4225] [Q&A]
I'm familiar with breaking up repetitive loop calculations into subsets and doing them with std::async  and std::future. I was thinking this might be a good thing to do with taxonomy::piecewise_function::evaluate(). 

@aothms what do you think?

Is there a preferred approach to multithreading in IFCOS?

**Comment by aothms** [comment]
We have a bit of a threadpool using std::async here

https://github.com/IfcOpenShell/IfcOpenShell/blob/f12c3aae09f0bb8a981ffdab19a39c2b32a6e5ed/src/ifcgeom/Iterator.h#L307-L317

But invoking the mapping functions is actually single threaded (because previously it didn't really do calculations; interfaces with IfcFile, which is threadsafe but not really parallel).

https://github.com/IfcOpenShell/IfcOpenShell/blob/f12c3aae09f0bb8a981ffdab19a39c2b32a6e5ed/src/ifcgeom/Iterator.h#L177

So, I don't know, I find it a bit hard to oversee. On the one hand if we do multi-threading on multiple levels (processing different alignments in parallel, but also the segments of a piecewise fn in parallel) we're likely just adding unnecessary context switches. But I'm not sure whether that will be the case now or in the future.

I'm also not sure if once we're doing various sweeps off of alignments whether the piecewise fn evaluation is actually significant or whether there are other performance optimisations we should look at first. Also, the code is already fairly complex.

This request has came in earlier also though. Also in the building bim models there can be a huge single product where it makes sense to parallelize parts of that product. Maybe coroutines are the answer? https://en.cppreference.com/w/cpp/language/coroutines

**Comment by RickBrice** [comment]
I'll just wait on this until there are alignment-related performance issues. Thanks for your insights.
